dhis2/dhis2-core
# fix: add short name column to Category, DataElementGroupSet, OrganisationUnitGroupSet [DHIS2-8937]

**Edit:** below is only for `Category` but we decided to also add `shortName` to `DataElementGroupSet` and `OrganisationUnitGroupSet`.

### Summary
This PR adds the `shortname` column to the `dataelementcategory` table (`Category` entity).
So far the `shortName` property was derived from `name` which caused issues when e.g. trying to PATCH the property (as reported in DHIS2-8937).

To make the added `shortname` column unique and not null the name is used to derive a unique short name.
This is done in 3 steps:

1. add new `shortname` column (SQL script)
2. fill all missing short names based on the name where unique names are found by adding a number to the end (JDBC script)
3. making the `shortname` column `not null` and `unique` (SQL script)

### Automatic Testing
The complex change was to come up with a unique name based on a given name that fits within the limit of maximum of 50 characters for the `shortname`.
A helper function was added `UniqueUtils#addUnique` which is unit tested by `UniqueUtilsTest`.

Many tests needed update of the test setup so the `shortName` is not null.

### Manual Testing
To properly test the DB migration the used database needs to contain categories with names that would clash in case one simply uses the first 50 characters.
The easiest way is to manually create a set of categories with names where the first 50 characters are identical.

1. (prepare your DB as described above)
2. start the server with the code changes of this PR
3. verify the `shortname` column was added using some DB tool
4. verify the `shortname` column is `not null` using some DB tool
5. verify the `shortname` column is `unqiue` using some DB tool

```sql
SELECT con.*
       FROM pg_catalog.pg_constraint con
            INNER JOIN pg_catalog.pg_class rel
                       ON rel.oid = con.conrelid
            INNER JOIN pg_catalog.pg_namespace nsp
                       ON nsp.oid = connamespace
       WHERE rel.relname = 'dataelementcategory';
```

When DB migration was successful we now also have to test that PATCHing the `shortName` property now works.

1. use GET http://localhost:8080/api/categories/ to find a category UID to update
2. use PATCH http://localhost:8080/api/categories/{UID}/shortName with payload `{"shortName": "new name"}`
3. use GET again to verify the short name was updated
